Astro results using PyCBC

vaibhavtewari opened the pull request:
I am opening a pull request for code that estimates rates for the two canonical mass distributions. It takes in prior from a previous run, uses injection files to estimate VT and estimate rate posteriors.

spxiwh commented:
Hi @vaibhavtewari thanks for posting this. Do you have a suggestion for who to review this? Maybe @tdent ? Before assigning someone to actually look over the code, some bookkeeping changes are needed:

 * Calling this `astro` is not particularly meaningful. Can you use a more meaningful name for this module.
 * pycbc_astro.py is an executable, can you move this to `bin` (in a new subfolder if relevant), remove the extension, add it to `setup.py` 
 * If this is a new module, I guess it would need an `__init__.py` to be consistent with the other modules (unless that's a policy we want to change?)
 * Please look over the codebeat comments, if there is anything useful there please act on it, but we don't require that codebeat is happy.
 * Please fix the codacy problems. We do ask that problems reported by codacy or landscape (if that will ever run again) get fixed.

ahnitz commented:
@vaibhavtewari Thanks for posting this! I think this gives us a much better how to integrate it. I agree with @spxiwh's suggestions. Also, as part of this, it would be useful to update the docstring documentation to match the numpydoc style. 

There are instructions here, but you cal also see examples in other parts of the package. 

https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

tdent commented:
It looks like travis is failing with 
`ImportError: No module named pystan` 
@ahnitz any idea about this - should pystan be given as a requirement in the package somewhere else or does travis need to be told about it some other way?

ahnitz commented:
@tdent I think we'd have to add it as a requirement in setup.py.

I wouldn't too much about those errors yet though until a bit further polish is applied. 

A couple things we should think about
1) check for PEP8 compliance
2) documentation
3) merge with rates.py and sensitivity.py??

A simple that you can do easily I think @vaibhavtewari 
1) Use the conversions module to get the mchirp etc conversions

tdent commented:
@vaibhavtewari @ahnitz I think we will also want the executable to take in a specified list of files rather than folder locations which could (in principle) contain anything.

vaibhavtewari commented:
Hi Ian, Alex, Tom

I applied the first set of corrections based on your feedback. Thank you for them. I can now run the analysis from within the pycbc. Tom is much aware about this analysis, so he will be a good reviewer, if he is available. 

Tom we can provide any delimiter. This will eventually be useful as it will choose the required files among many from the hdf_inj folder. Alex mchirp function can use the already in circulation one. Will keep this is mind as this progresses.

Please have a look and I will improve it further.

tdent commented:
@vaibhavtewari Please add options to specify a list of input injection file locations (I assume this means hdfinjfind files?), to specify the bank file location, and to specify the statmap file location(s), and make these mutually exclusive with the folder-based input arguments.  I'm not sure what you mean by delimiter, can you give an example? 
Also address the Codacy issues - 'yellow' and 'red' ones have priority.

ahnitz commented on pycbc/population/scale_injections.py:
```diff
@@ -0,0 +1,796 @@
+   
+    return injs
+        
+def m1m2_to_mcheta(m1, m2):
```
@vaibhavtewari Can you replace these types of functions with the equivelant in pycbc.conversions? Or if you have some new ones, please make a PR with just that change and add it to the pycbc.conversions module.

vaibhavtewari replied:
I am just using m1m2->mch conversion. Now using from pycbc.conversion

ahnitz commented on pycbc/population/rates_functions.py:
```diff
@@ -0,0 +1,443 @@
+    return np.resize(m1, nsamples), np.resize(m2, nsamples)
+
+def m1m2_to_mcheta(m1, m2):
+    ''' Get chirp mass and eta for m1 and m2
```
Please use double quotes here.

vaibhavtewari replied:
This function has been replaced by one in pycbc.conversions

ahnitz commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,216 @@
+#                    version=pycbc.version.git_verbose_msg)
+#group = parser.add_mutually_exclusive_group(required=True)
+
+parser.add_argument('--sim-folder', dest='sim_folder', required=True, 
```
Can you make the command line arguments take in the list of files it needs directly? Similar with the output produced. It should be explicit what files are being read and written.

vaibhavtewari replied:
Alex this is not clear to me. Could you please explain ?

ahnitz replied:
Sure, I'd suggest the program take in the files directly instead of a folder. So on the command line you have

--inj-file inj1.hdf inj2.hdf ....
and possibly other options for other files you need. 

The point is that it's a bit dangerous to expect files to have some specific naming convention or to be laid out in a specific way. Also workflow managers generally require that these conditions never be assumed.

ahnitz commented on pycbc/population/scale_injections.py:
```diff
@@ -0,0 +1,796 @@
+            return (np.log(xx) - np.log(min_m2))/(np.log(max_m2) - np.log(min_m2))        
+    
+        p_thr = 0.0001
+        stat, pvalue = kstest(mass1, cdf_componentMass)
```
Why are you doing a statistical test to determine which distribution it was? Why not simply take this as a command line option? (or keyword to the explicit functions).

vaibhavtewari replied:
This was to make things least dependent on user. @tdent we need all the BBH files in all the folders, the delimiter can find files in multiple folders e.g. '*BBH*/*hdfinj*.hdf" and the statistical test will identify the injection distribution with nice accuracy. Alternatively we can specify every single BBH file and the corresponding injection strategy used.

vaibhavtewari replied:
*BBH*/*hdfinj*.hdf

ahnitz commented on pycbc/population/scale_injections.py:
```diff
@@ -0,0 +1,796 @@
+        
+    return m1, m2
+
+def jacobian_m1m2_to_mcheta(mch, eta, minm1, maxm1):
```
We have jacobians built into the distributions / transformations we also use for pycbc inference. Can you coordinate with @cdcapano  to use that functionality to avoid the duplication.

vaibhavtewari replied:
I have removed the function. It was meant if the PDF of astro model was defined on the mch-eta space. Will bring it/Collin's version back if required in future.

vaibhavtewari commented:
Hi All, Could some one please help this get reviewed ? It should not take a lot of time. I will be more than happy to produce various sanity checks.

spxiwh commented:
This should be rebased against master at this point

titodalcanton commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,237 @@
+import matplotlib as mpl
+
+# Parse command line
+parser = argparse.ArgumentParser(description='Estimate rates for flat in log and power-law mass distribution.')
```
@vaibhavtewari a minor fix here to avoid future confusion: could you improve the docstring at the top of the file (the current one seems out of date or incomplete) and then replace this line with:
```
parser = argparse.ArgumentParser(description=__doc__)
```

vaibhavtewari replied:
@titodalcanton this is unclear to me. The current docstring for description is 
parser = argparse.ArgumentParser(description = "Estimate rates for flat in "
                                      "log and power-law mass distribution.")

Are you talking about something else ?

vaibhavtewari commented:
Hi all,  thanks for your feedback. I think I rebased the pull but it is still failing at importing pystan. Is that right ?

tdent commented:
@vaibhavtewari yes, no pystan is one error.  There also seems to be a warning-treated-as-error since the documentation build (I think) can't find any RST file for the population module - see https://travis-ci.org/ligo-cbc/pycbc/jobs/330787493#L3869 .  That should be easy to fix though.

ahnitz commented:
@vaibhavtewari Out of curiosity what is the reason we need to use pystan? If there's a straightforward way to avoid a new dependency that's usually something good to think about. Otherwise, we can add it.

cdcapano commented on pycbc/population/rates_functions.py:
```diff
@@ -0,0 +1,424 @@
+    ff, flag = optimize.leastsq(optm, [0.1, mu_norm, sigma_norm], args=(p_xs, xs))
+    return ff[0], ff[1], ff[2]
+
+# The flat in log and power-law mass distribution models  #
```
I'd prefer if you used the distributions that are defined in the distributions module for the following instead. There, [https://github.com/ligo-cbc/pycbc/blob/master/pycbc/distributions/uniform_log.py#L22](uniform log) and [https://github.com/ligo-cbc/pycbc/blob/master/pycbc/distributions/power_law.py#L23](power law) distributions are defined. Do they do what you need? If not, can you clarify what else they would need? I'd prefer to add any addition functionality you need to those classes if possible.

vaibhavtewari replied:
Thanks for the suggestions. I would personally prefer to keep astrophysical distributions within the population module. It might be better to add one more file models.py and populate it as we move forward in number of detections. We can talk more about this.

tdent commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,237 @@
+           'alpha': alpha[dist],
+           'mu': mu[dist],
+           'sigma': vt_err_w_cal_uncrt[dist]}
+    rate_fit = pystan.stan(file=opts.stan_file, data=rate_data[dist], iter=20000)
```
FYI @cdcapano : this is directly from the reviewed O1 FGMC rates analysis written by Will Farr.  A .stan file is the specification of the statistical model that is being sampled, it would be order(50-100) lines so very impractical to include as such in the help message.  The call to pystan does the MC sampling of the model.  To continue doing what the O1 rates analysis did, it's necessary to continue using pystan *or* to replace it with some other statistical modelling and sampling package - but to review that (as yet non-existent) alternative - which would have to be coded up by somebody using another sampling module - would require showing it did the same as pystan, so in either case we need pystan.  Let's get the O1 analysis as-is into the pipeline first and worry about cleanliness of imports later.

vaibhavtewari replied:
I am not using pystan anymore, but have introduced a function fg_mc in the file rates_functions.py, please have a look

vaibhavtewari commented:
Hi all, thanks very much for the feedback. I will implement them soon. @ahnitz @tdent @cdcapano I managed to do FGMC using basic tools so we dont need pystan for now. Will implement it today.

ahnitz commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,237 @@
+
+p_astro = [1 + expm1(np.sort(log_pastros_imf)[::-1]), 1 + expm1(np.sort(log_pastros_lnm)[::-1])]
+   
+print p_astro 
```
In any case, if you must output it to the screen either use logging module if this is optional, or use the print function. One less thing to change when we move to python3.

cdcapano commented:
@vaibhavtewari Seeing @tdent's comment about pystan, I didn't realize that had been reviewed for the O1 analysis. In that case, I'd say stick with pystan rather than try to redo it, but that will require adding pystan to the requirements. Not sure how @ahnitz feels about that...

ahnitz commented:
I think if Vaibhav already has something close to working with basic tools,
that's quite nice actually. It in principle, should be easy to cross check
with the original version.

On Tue, Jan 23, 2018 at 5:16 PM, Collin Capano <notifications@github.com>
wrote:

> @vaibhavtewari <https://github.com/vaibhavtewari> Seeing @tdent
> <https://github.com/tdent>'s comment about pystan, I didn't realize that
> had been reviewed for the O1 analysis. In that case, I'd say stick with
> pystan rather than try to redo it, but that will require adding pystan to
> the requirements. Not sure how @ahnitz <https://github.com/ahnitz> feels
> about that...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ligo-cbc/pycbc/pull/2009#issuecomment-359843077>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACGrRqOPnsQTgPpNUFNIBdkcdmneoTmeks5tNgXzgaJpZM4Q52TI>
> .
>



-- 
Dr. Alexander Nitz
Max Planck Institute for Gravitational Physics (Albert Einstein Institute)
Callinstrasse 38
D-30167 Hannover,  Germany
Tel: +49 511 762-17097

vaibhavtewari commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,237 @@
+parser.add_argument('--bank-folder', dest='bank_folder', required=True, 
+                  help='Folder containing simulation files - delimiters can be used.')
+parser.add_argument('--statmap-folder', dest='statmap_folder', required=True,
+                  help="Folder containing the STATMAP files (files that save background and foregound trigger information.")
```
done for the most part!

vaibhavtewari commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,237 @@
+#    width, center = (bins[1] - bins[0]), (bins[:-1] + bins[1:]) / 2
+#    plt.bar(center, hist4, align='center', width=width, alpha = 0.8, label = label[dist])
+#    plt.legend(loc = 'best')
+#plt.title('SNR Fall Off') 
```
done

vaibhavtewari commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,237 @@
+fg_bins = np.logspace(np.log10(opts.thr_val), np.log10(np.max(fg_stats)), 101)
+fg_stats = fg_stats[fg_stats > opts.thr_val]
+
+ts = np.linspace(opts.thr_val, 11.0, 1000)
```
This was for a plot which has been removed.

vaibhavtewari commented on bin/population/pycbc_population_rates:
```diff
@@ -0,0 +1,237 @@
+
+#Estimate p_astro for top 5 events
+
+from numpy import logaddexp, log, newaxis, expm1
```
done

vaibhavtewari commented on pycbc/population/rates_functions.py:
```diff
@@ -0,0 +1,424 @@
+
+    def skew(x, alpha, mu, sigma):
+        t = (x - mu) / sigma
+        return 2 / sigma * pdf(t) * cdf(alpha*t)
```
Unfortunately this seems unavoidable under the current method. However, reduced 3 functions to 1. The function is called only once.

vaibhavtewari commented on pycbc/population/rates_functions.py:
```diff
@@ -0,0 +1,424 @@
+
+    def optm(l, y, x):
+            alpha, mu, sigma = l
+            return kde(x)- skew(x, alpha, mu, sigma)    
```
No fixed this. Thanks for the suggestion.

vaibhavtewari commented on pycbc/population/rates_functions.py:
```diff
@@ -0,0 +1,424 @@
+    kde = ss.gaussian_kde(R)
+    p_xs = kde(xs)
+    
+    ff, flag = optimize.leastsq(optm, [0.1, mu_norm, sigma_norm], args=(p_xs, xs))
```
Made it much better.

vaibhavtewari commented:
Hi Collin,
Thanks very much for the review. All the feedback was very helpful. I have incorporated all the suggestions except one, which I will do soon.

spxiwh commented:
Here's landscape's review on this:

https://landscape.io/github/spxiwh/pycbc/472/diff/fixed?page=1#errors

Some of these (unused variables, or documentation strings being used in place of comments) are harmless, but some of the other issues are potentially problematic.

tdent commented on pycbc/population/rates_functions.py:
```diff
@@ -0,0 +1,478 @@
+    Function to fit the likelihood Fixme
+    '''
+
+    def log_sum_exp(u, v):
```
can numpy.logaddexp be used here?

tdent replied:
@vaibhavtewari this question is still open - is this custom function needed?

ahnitz commented:
@vaibhavtewari The error from Travis is this. 

"""

Warning, treated as error:
/home/travis/build/pycbc-build/environment/lib/python2.7/site-packages/PyCBC-0e7919-py2.7.egg/pycbc/population/scale_injections.py:docstring of pycbc.population.scale_injections.estimate_vt:5:Unexpected section title.

Parameters
----------
make: *** [html] Error 1
Traceback (most recent call last):

"""

ahnitz commented on pycbc/population/scale_injections.py:
```diff
@@ -0,0 +1,716 @@
+    #Try including ifar threshold
+    '''Based on injection strategy and the desired astro model estimate the injected volume.
+       Scale injections and estimate sensitive volume.
+a
```
I think this stray character might be causing the problem?

vaibhavtewari commented:
@ahnitz Whats the reason for the error ?
Warning, treated as error:
/home/travis/build/ligo-cbc/pycbc/docs/pycbc.population.rst:document isn't included in any toctree
make: *** [html] Error 1

ahnitz commented:
@spxiwh Can you take a look?

spxiwh commented:
That error implies that you've added a documentation page, but there's no link to it. It must be possible to reach documentation pages from link(s) off of the home page.

vaibhavtewari commented:
Hi Ian, Sorry I dont understand what you said. Could you please explain ? I am getting the same error when I build the documentation from the terminal. So If I am not suceeding to build documentation, how can I have a documentation page ?

spxiwh commented:
[After having a closer look at what's going on here]: The error indicated that the module you added was not being linked in the auto-generated page here:

http://pycbc.org/pycbc/latest/html/py-modindex.html

I don't know why it was failing, but restarting Travis appears to have resolved the problem. Perhaps some files were being cached that should have been regenerated within the documentation build test? Anyway, problem has disappeared now and Travis is building.

vaibhavtewari commented:
Thanks Ian. The build passed :)

vaibhavtewari commented:
Not sure what is the next step, but this is what we could publish on the astro page summary
1) The rate prior/posterior plot for two canonical models
2) The 5-95 percentile rates 
3) The p_astro value for the top events
Any comments on this are wlecome.

The code is doing all that I want to understand what is the best way to pass this information.

ahnitz commented:
@vaibhavtewari I think it would be good to add features in separate (hopefully much smaller PRs that don't require so much review, sorry about this making so much work for you!)

@spxiwh @cdcapano Any chance you can take another pass through about your concerns? 

@vaibhavtewari Are there any comments you haven't addressed yet?

tdent commented:
@vaibhavtewari I think if all the review and code quality issues here have been resolved then the request should be merged and then we can check that the code functions in the workflow and discuss what sort of outputs we want in a new issue (or on chat / in a telecon).  As Alex says we shouldn't load more things into one request than necessary.

spxiwh commented:
I would like to see the issues raised by landscape resolved (or responded to). As I said above, some of these look dangerous:

https://landscape.io/github/spxiwh/pycbc/472/diff/fixed?page=1#errors

Agree that this should be merged now, and new features *not* added to this pull request.

spxiwh commented:
@tdent suggested that most of the landscape issues have been fixed. Here's an updated landscape report:

https://landscape.io/github/spxiwh/pycbc/497/diff

Indeed now the only problems are unused variables. I think the bottom one should be fixed, and I leave the other ones up to you.

vaibhavtewari commented:
Hi all, thanks very much for all the feedback. 
@ahnitz Sorry I didnt understand yoru comment on smaller PRs. Is that meant for future ? I think I have addressed all the suggestions.
@Ian I left the spin variables in the functions as it is, in case we want to include them in future. Fixed the last duggestion on landscape.
@tdent fg_mc is a replacement for pystan. It does things in not the most ideal fashion but does it faster and better than pystan. It iterates over Bayesian equation for rates recursively until the prior and posterior on background events are very close. The prior on rates is provided by the user. It does that by slicing starting with a very large number of samples of prior and slices at each cycle. Do you think you can have a look ? I see the logic and output are fine but an addition review will be great. It does run the risk of running out of samples to slice if posterior is very different from prior. But then it is a problem in itself.

tdent commented:
@vaibhavtewari by the way, did you look at my comment about the 'log_sum_exp' function, is it needed? 
For the functionality of the fg_mc calculation itself I would need some time to think about what it is doing.  I think in this pull request the best thing to do is add some more documentation to the fg_mc code, specifically numpy-style descriptions of inputs and outputs and an overview of what the calculation doing.

vaibhavtewari commented:
Yeah Tom log_sum_esp is needed. Its a pystan function, I just recoded it.

tdent commented:
@vaibhavtewari Can you check that the numpy logaddexp function doesn't actually do the same thing?  I think the function that was written in pystan was just another way of writing log(exp(u) + exp(v)) since pystan might not have a built in function.  I.e. can you test whether this function gives a different result from numpy logaddexp.

vaibhavtewari commented:
@tdent you are right. Now using numpy function.

tdent commented:
I don't know why travis is failing again (@spxiwh any idea?)- but assuming that this doesn't indicate a problem with the pull request, then I think it is ready to be merged now.

spxiwh commented:
@vaibhavtewari The empty `__init__.py` file in the new module seems to be causing some problems with the auto-documentation. I think this might actually be a bug in Sphinx, but can you make the following change:

https://github.com/ligo-cbc/pycbc/pull/2075/commits/805b0f27c5d8c6f8247b3a6315c4c8fd46f11cdc

to the new `__init__.py` file. This is then consistent with other PyCBC modules, and resolves the issue with the documentation (which is currently causing Travis to fail).

vaibhavtewari commented:
Hi @ahnitz @spxiwh this can be merged now. I will open a separate PR as I intend to split a script. But the current version has been approved by 3 reviewers.

tdent commented:
@spxiwh @vaibhavtewari Codacy doesn't seem to like the new imports in the most recent commit (it says the imports are unused) - is that nevertheless OK to proceed with merging?

spxiwh commented:
I see no issues with merging this from my perspective (there is a general style question about whether we are using `__init__` files in the right way, but that's outside the scope of this patch).

Given that this has taken a while, there are 3 approving reviewers, and no issues from the automated checks, I'm pushing the merge button.
